nav2_waypoint_follower/src/waypoint_follower.cpp

```diff
@@ -0,0 +1,224 @@
+      goal = action_server_->accept_pending_goal();
+      goal_index = 0;
+      new_goal = true;
+    }
+
+    // Check if we need to send a new goal
+    if (new_goal) {
+      new_goal = false;
+      ClientT::Goal client_goal;
+      client_goal.pose = goal->poses[goal_index];
+
+      auto send_goal_options = rclcpp_action::Client<ClientT>::SendGoalOptions();
+      send_goal_options.result_callback =
+        std::bind(&WaypointFollower::resultCallback, this, std::placeholders::_1);
+      send_goal_options.goal_response_callback =
+        std::bind(&WaypointFollower::goalResponseCallback, this, std::placeholders::_1);
+      auto future_goal_handle =
+        nav_to_pose_client_->async_send_goal(client_goal, send_goal_options);
+      current_goal_status_ = ActionStatus::PROCESSING;
+    }
+
+    feedback->current_waypoint = goal_index;
+    action_server_->publish_feedback(feedback);
+
+    if (current_goal_status_ == ActionStatus::FAILED) {
+      if (stop_on_failure_) {
+        RCLCPP_WARN(get_logger(), "Failed to process waypoint %i in waypoint "
+          "list and stop on failure is enabled."
+          " Terminating action.", goal_index);
+        action_server_->terminate_goals();
+        return;
+      } else {
+        RCLCPP_INFO(get_logger(), "Failed to process waypoint %i,"
+          " moving to next.", goal_index);
+      }
+    } else if (current_goal_status_ == ActionStatus::SUCCEEDED) {
+      RCLCPP_INFO(get_logger(), "Succeeded processing waypoint %i, "
+        "moving to next.", goal_index);
+    } else if (current_goal_status_ != ActionStatus::PROCESSING) {
+      // Update server state
+      goal_index++;
+      new_goal = true;
+      if (goal_index >= goal->poses.size()) {
+        RCLCPP_INFO(get_logger(), "Completed all %i waypoints requested.",
+          goal->poses.size());
+        action_server_->succeeded_current(result);
+      }
+    } else {
+      RCLCPP_DEBUG(get_logger(), "Processing waypoint %i...", goal_index);
+    }
+
+    r.sleep();
+  }
+}
+
+void
+WaypointFollower::resultCallback(
+  const rclcpp_action::ClientGoalHandle<ClientT>::WrappedResult & result)
+{
+  switch (result.code) {
+    case rclcpp_action::ResultCode::SUCCEEDED:
+      current_goal_status_ = ActionStatus::SUCCEEDED;
+      return;
+    case rclcpp_action::ResultCode::ABORTED:
+      current_goal_status_ = ActionStatus::FAILED;
+      return;
+    case rclcpp_action::ResultCode::CANCELED:
+      current_goal_status_ = ActionStatus::FAILED;
+      return;
+    default:
+      current_goal_status_ = ActionStatus::UNKNOWN;
+      return;
+  }
+}
+
+void
+WaypointFollower::goalResponseCallback(
+  std::shared_future<rclcpp_action::ClientGoalHandle<ClientT>::SharedPtr> future)
+{
+  auto goal_handle = future.get();
+  if (!goal_handle) {
+    RCLCPP_ERROR(get_logger(),
+      "NavigateToPose client failed to send goal to server.");
+    current_goal_status_ = ActionStatus::FAILED;
+  }
```

Comment:
We also need to check if the goal was rejected.

Reply:
I’m fairly certain that’s how this works- I checked elsewhere in the codebase and everywhere uses this check for rejection. This also includes the rclcpp action client tutorials

Reply:
OK.

On the separate topic, agree we should have a discussion about the simple action server and also incorporate recent [feedback](https://github.com/ros2/rclcpp/issues/759).